### PR TITLE
[SLP]Improve minbitwidth analysis for shifts.

### PR DESCRIPTION
**Pull request by alexey-bataev**
Adds improved bitwidth analysis for shl/ashr/lshr instructions. The
analysis is based on similar version in InstCombiner.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -14039,6 +14039,84 @@ bool BoUpSLP::collectValuesToDemote(
...
+    // smaller ashr iff we know that all the bits from the sign bit of the
+    // original type and the sign bit of the truncate type are similar.
+    uint32_t OrigBitWidth = DL->getTypeSizeInBits(V->getType());
+    unsigned Level1, Level2;
+    KnownBits AmtKnownBits = computeKnownBits(I->getOperand(1), *DL);
+    unsigned ShiftedBits = OrigBitWidth - BitWidth;
+    if (AmtKnownBits.getMaxValue().uge(BitWidth) ||
+        ShiftedBits >=
+            ComputeNumSignBits(I->getOperand(0), *DL, 0, AC, nullptr, DT))
+      return IsProfitableToDemote && IsPotentiallyTruncated(I, BitWidth);
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:14106]: If this section of each shift instruction type was pulled out into a small helper, then the 3 opcodes could share the rest of the code and we'd get rid of a lot of duplication - is it worth it?

**Comment by alexey-bataev** [comment]
Ping!

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by mikaelholmen** [comment]
Hi @alexey-bataev ,

I think we get a miscompile with this patch
Reproduce with:
```
opt bbi-93743_2.ll -mtriple=aarch64 -passes=slp-vectorizer -S -o - -debug -slp-threshold=-100
```
The problem occurs if both inputs to foo are 0xffffffffffffffff.

Then the
```
  %5 = shufflevector <2 x i128> %3, <2 x i128> %4, <2 x i32> <i32 0, i32 3>
```
in the slp-vectorizer output will result in
```
<0x10000000000000000,poison>
```
due to the "nsw" on the "shl", and that poison then turns the return value from the function to be poison as well.

[bbi-93743_2.ll.gz](https://github.com/llvm/llvm-project/files/14756140/bbi-93743_2.ll.gz)

**Comment by alexey-bataev** [comment]
Thanks, will fix it ASAP

**Comment by alexey-bataev** [comment]
> Hi @alexey-bataev ,
> 
> I think we get a miscompile with this patch Reproduce with:
> 
> ```
> opt bbi-93743_2.ll -mtriple=aarch64 -passes=slp-vectorizer -S -o - -slp-threshold=-100
> ```
> 
> The problem occurs if both inputs to foo are 0xffffffffffffffff.
> 
> Then the
> 
> ```
>   %5 = shufflevector <2 x i128> %3, <2 x i128> %4, <2 x i32> <i32 0, i32 3>
> ```
> 
> in the slp-vectorizer output will result in
> 
> ```
> <0x10000000000000000,poison>
> ```
> 
> due to the "nsw" on the "shl", and that poison then turns the return value from the function to be poison as well.
> 
> [bbi-93743_2.ll.gz](https://github.com/llvm/llvm-project/files/14756140/bbi-93743_2.ll.gz)

Ok, it is not an issue of this patch, but a long-standing issue in vectorizer. I haму a fix got this in one of the extra patches, will commit it separately with your reproducer.

**Comment by alexey-bataev** [comment]
> Hi @alexey-bataev ,
> 
> I think we get a miscompile with this patch Reproduce with:
> 
> ```
> opt bbi-93743_2.ll -mtriple=aarch64 -passes=slp-vectorizer -S -o - -slp-threshold=-100
> ```
> 
> The problem occurs if both inputs to foo are 0xffffffffffffffff.
> 
> Then the
> 
> ```
>   %5 = shufflevector <2 x i128> %3, <2 x i128> %4, <2 x i32> <i32 0, i32 3>
> ```
> 
> in the slp-vectorizer output will result in
> 
> ```
> <0x10000000000000000,poison>
> ```
> 
> due to the "nsw" on the "shl", and that poison then turns the return value from the function to be poison as well.
> 
> [bbi-93743_2.ll.gz](https://github.com/llvm/llvm-project/files/14756140/bbi-93743_2.ll.gz)



> Hi @alexey-bataev ,
> 
> I think we get a miscompile with this patch Reproduce with:
> 
> ```
> opt bbi-93743_2.ll -mtriple=aarch64 -passes=slp-vectorizer -S -o - -slp-threshold=-100
> ```
> 
> The problem occurs if both inputs to foo are 0xffffffffffffffff.
> 
> Then the
> 
> ```
>   %5 = shufflevector <2 x i128> %3, <2 x i128> %4, <2 x i32> <i32 0, i32 3>
> ```
> 
> in the slp-vectorizer output will result in
> 
> ```
> <0x10000000000000000,poison>
> ```
> 
> due to the "nsw" on the "shl", and that poison then turns the return value from the function to be poison as well.
> 
> [bbi-93743_2.ll.gz](https://github.com/llvm/llvm-project/files/14756140/bbi-93743_2.ll.gz)

Fixed in 26dd12871ca8a3dcc5e6920add231af354310088

**Comment by mikaelholmen** [comment]
> Fixed in [26dd128](https://github.com/llvm/llvm-project/commit/26dd12871ca8a3dcc5e6920add231af354310088)

I verified that the fix solves the problem I saw. Thanks!

**Pull request merged**